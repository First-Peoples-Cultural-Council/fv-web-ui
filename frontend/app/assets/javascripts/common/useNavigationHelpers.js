import useWindowPath from 'DataSource/useWindowPath'
import NavigationHelpers, { getNewPaginationUrl } from 'common/NavigationHelpers'
import URLHelpers from 'common/URLHelpers'
/**
 * @summary useNavigationHelpers
 * @description Custom hook to make working with NavigationHelpers easier
 *
 * @version 1.0.1
 *
 * @component
 *
 * @returns object {object} {changePagination}
 * @returns object.changePagination {function} Url is updated when changePagination({page, pageSize}) is called
 */
function useNavigationHelpers() {
  const { splitWindowPath, pushWindowPath } = useWindowPath()
  return {
    changePagination: ({ page, pageSize }) => {
      NavigationHelpers.navigate(getNewPaginationUrl({ splitWindowPath, page, pageSize }), pushWindowPath, false)
    },
    navigate: (url) => {
      NavigationHelpers.navigate(url, pushWindowPath, false)
    },
    getBaseURL: () => {
      return URLHelpers.getBaseURL()
    },
  }
}

export default useNavigationHelpers
