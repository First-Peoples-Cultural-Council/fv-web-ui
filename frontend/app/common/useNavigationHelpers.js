import useWindowPath from 'dataSources/useWindowPath'
import NavigationHelpers, { getNewPaginationUrl } from 'common/NavigationHelpers'
import URLHelpers from 'common/URLHelpers'
/**
 * @summary useNavigationHelpers
 * @description Custom hook to make working with NavigationHelpers easier
 *
 * @version 1.0.1
 *
 * @component
 *
 * @returns object {object} {changePagination}
 * @returns object.changePagination {function} Url is updated when changePagination({page, pageSize}) is called
 */
function useNavigationHelpers() {
  const { splitWindowPath, pushWindowPath, replaceWindowPath } = useWindowPath()
  return {
    changePagination: ({ page, pageSize }) => {
      NavigationHelpers.navigate(getNewPaginationUrl({ splitWindowPath, page, pageSize }), pushWindowPath, false)
    },
    navigate: (url) => {
      NavigationHelpers.navigate(url, pushWindowPath, false)
    },
    navigateReplace: (url) => {
      NavigationHelpers.navigate(url, replaceWindowPath, false)
    },
    getSearchObject: () => {
      if (window.location.search === '') {
        return {}
      }

      const search = {}
      const searchParams = (window.location.search || '?').replace(/^\?/, '')
      searchParams.split('&').forEach((item) => {
        if (item !== '' && /=/.test(item)) {
          const propValue = item.split('=')
          search[propValue[0]] = propValue[1]
        }
      })
      return search
    },
    getBaseURL: () => {
      return URLHelpers.getBaseURL()
    },
  }
}

export default useNavigationHelpers
