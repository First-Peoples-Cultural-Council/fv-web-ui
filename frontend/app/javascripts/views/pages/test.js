/*
Copyright 2016 First People's Cultural Council

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Immutable from 'immutable'

// REDUX
import { connect } from 'react-redux'
// REDUX: actions/dispatch/func
import { fetchPortal } from 'reducers/fvPortal'
import { fetchSharedPictures } from 'reducers/fvPicture'
import { fetchWord } from 'reducers/fvWord'
import { pushWindowPath } from 'reducers/windowPath'

import PromiseWrapper from 'componentsShared/PromiseWrapper'

const DefaultFetcherParams = { filters: { 'properties.dc:title': '', dialect: '78086057-9c34-48f7-995f-9dc3b313231b' } }

/**
 * Explore Archive page shows all the families in the archive
 */

const { func, object } = PropTypes
export class Test extends Component {
  static propTypes = {
    routeParams: object.isRequired,
    // REDUX: reducers/state
    computePortal: object.isRequired,
    computeSharedPictures: object.isRequired,
    computeWord: object.isRequired,
    properties: object.isRequired,
    // REDUX: actions/dispatch/func
    fetchPortal: func.isRequired,
    fetchSharedPictures: func.isRequired,
    fetchWord: func.isRequired,
    pushWindowPath: func.isRequired,
  }

  constructor(props, context) {
    super(props, context)

    /*let computeEntities = new List();

    let computePortal = new Map([['id', '/FV/sections/Data/TestFamily/TestLanguage/PopoDialect'], ['entity', this.props.computePortal]]);
*/

    this.state = {
      filteredList: null,
      fetcherParams: Object.assign(
        {
          currentPageIndex: 0,
          pageSize: 10,
        },
        DefaultFetcherParams
      ),
      pathOrId: '/FV/Workspaces/Data/Athabascan/Beaver/Tsaa î%20Dane%20-%20Beaver%20People/Dictionary/119430',
    }

    this.fetchData = this.fetchData.bind(this)
    this.fixedListFetcher = this.fixedListFetcher.bind(this)

    /*var client = new elasticsearch.Client({
      host: 'https://preprod.firstvoices.com/nuxeo/site/es',
      httpAuth: 'null:null',
      log: 'trace'
    });

    var test123 = client.search({
      index: 'nuxeo',
      'enrichers.document': 'breadcrumb', // enrichers -- not working here
      'enrichers.document': 'breadcrumb', // enrichers -- not working here
      'fetch.document' : 'dc:creator', // marshallers - not working here
      headers: {
        'enrichers.document': 'breadcrumb', // enrichers -- not working here
        'enrichers.document': 'breadcrumb', // enrichers -- not working here
        'fetch.document' : 'dc:creator' // marshallers - not working here
      },
      body: {
          "from" : 0,
          "size" : 100,
          //"_source": ["dc:title"],
          "query": {
            "filtered": {
              "filter": {
                  "bool": {
                      "must":
                          [
                              {
                                "term": {
                                  "ecm:primaryType": "FVPortal"
                                }
                              },
                              {
                                "term": {
                                  "ecm:isVersion": 0
                                }
                              },
                              {
                                "prefix": {
                                  "ecm:path": "/FV/Workspaces/"
                                }
                              }
                          ]
                      ,
                      "must_not": [
                        {"term": {
                          "ecm:isTrashed": "0"
                        }}
                      ]
                    }
              }
            }
          }
      }
    }).then(function (resp) {
        var hits = resp.hits.hits;
    }, function (err) {
        console.trace(err.message);
    });*/
  }

  fixedListFetcher(list) {
    this.setState({
      filteredList: list,
    })
  }

  fetchData() {
    this.props.fetchWord(this.state.pathOrId)
  }

  // Fetch data on initial render
  componentDidMount() {
    this.fetchData()
  }

  // Refetch data on URL change
  componentWillReceiveProps(/*nextProps*/) {
    //if (nextProps.windowPath !== this.props.windowPath || nextProps.routeParams.area != this.props.routeParams.area) {
    //  this.fetchData(nextProps);
    //}
  }

  /*shouldComponentUpdate(nextProps, nextState) {
    console.log(nextProps.computePortal.size);
    if (nextProps.computePortal.size === 0) {
      return false;
    }

    return true;
  }*/

  _onNavigateRequest() {}

  render() {
    //let computeEntities = new List();
    //let computePortal = new Map([['id', '/FV/sections/Data/TestFamily/TestLanguage/PopoDialect'], ['entity', this.props.computePortal]]);

    const computeEntities = Immutable.fromJS([
      {
        id: this.state.pathOrId,
        entity: this.props.computeWord,
      },
    ])

    /*let portalOperation = ProviderHelpers.getEntry(computePortal, '/FV/sections/Data/TestFamily/TestLanguage/PopoDialect');
    console.log(portalOperation);


    if (!portalOperation || portalOperation.isFetching) {
      return <div><CircularProgress mode="indeterminate" size={5} /> {selectn('message', portalOperation)}</div>;
    }

    if (portalOperation.isError) {
      return <div>{portalOperation.message}</div>
    }

    let portalResponse = selectn('response', portalOperation);*/

    return (
      <div className="row">
        <pre>
          <PromiseWrapper computeEntities={computeEntities}>
            {JSON.stringify(this.props.computeWord, null, '\t')}
          </PromiseWrapper>
        </pre>
      </div>
    )
  }
}

// REDUX: reducers/state
const mapStateToProps = (state /*, ownProps*/) => {
  const { fvPicture, fvPortal, fvWord, navigation } = state

  const { properties } = navigation
  const { computePortal } = fvPortal
  const { computeSharedPictures } = fvPicture
  const { computeWord } = fvWord

  return {
    computePortal,
    computeSharedPictures,
    computeWord,
    properties,
  }
}

// REDUX: actions/dispatch/func
const mapDispatchToProps = {
  fetchPortal,
  fetchSharedPictures,
  fetchWord,
  pushWindowPath,
}

export default connect(mapStateToProps, mapDispatchToProps)(Test)
