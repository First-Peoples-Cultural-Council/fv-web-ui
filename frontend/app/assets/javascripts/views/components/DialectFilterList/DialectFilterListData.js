/*
generateDataForHistoryEvents: this generates data used when the back/forward buttons are clicked.
May not be necessary if componentDidUpdate handles everything and/or can't we use data created by `generateListItemData`?
*/
import { Component } from 'react'
import { connect } from 'react-redux'
import { Set } from 'immutable'
import PropTypes from 'prop-types'
import selectn from 'selectn'
import { appendPathArrayAfterLandmark } from 'common/NavigationHelpers'

// REDUX: actions/dispatch/func
import { fetchCategories } from 'providers/redux/reducers/fvCategory'

import ProviderHelpers from 'common/ProviderHelpers'

class DialectFilterListData extends Component {
  historyData = {}
  filtersSorted = []
  selectedDialectFilter = undefined

  constructor(props) {
    super(props)

    const { workspaceKey } = this.props
    this.state = {
      // `switchWorkspaceSectionKeys` will return the related `section` "key" when passed a `Workspaces` "key". Think this value would be used in network/nql requests somewhere
      facetField: workspaceKey
        ? ProviderHelpers.switchWorkspaceSectionKeys(workspaceKey, this.props.routeParams.area)
        : '',
      facets: [],
    }
  }
  async componentDidMount() {
    // Bind history events
    window.addEventListener('popstate', this.handleHistoryEvent)

    const { facetType, facets } = this.props

    const selectedDialectFilter = selectn(`${facetType}`, this.props.routeParams)

    if (selectedDialectFilter) {
      this.selectedDialectFilter = selectedDialectFilter
    }

    // We may have data, update this.filtersSorted & history data
    if (facets && facets.length > 0) {
      this.filtersSorted = this.sortDialectFilters(facets)
      this.historyData = this.generateDataForHistoryEvents(this.filtersSorted)

      this.setState(
        {
          facets,
        },
        () => {
          if (this.selectedDialectFilter) {
            const selectedParams = this.historyData[this.selectedDialectFilter]
            if (selectedParams) {
              this.setSelected(selectedParams)
              this.selectedDialectFilter = undefined
            }
          }
        }
      )
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { facetType } = this.props

    let facets
    let prevFacets

    // Are we using facets or internal state?
    if (this.props.facets) {
      facets = this.props.facets
      prevFacets = prevProps.facets
    } else {
      facets = this.state.facets
      prevFacets = prevState.facets
    }
    // Note: guard against undefined
    const prevAppliedFilterIds = prevProps.appliedFilterIds || new Set()
    const currentAppliedFilterIds = this.props.appliedFilterIds || new Set()

    if (prevFacets.length !== facets.length) {
      this.filtersSorted = this.sortDialectFilters(facets)
    }

    if (prevFacets.length !== facets.length || prevAppliedFilterIds.equals(currentAppliedFilterIds) === false) {
      this.historyData = this.generateDataForHistoryEvents(this.filtersSorted)
    }

    // Is something selected? (via url)
    // NOTE: could this be driven by the `appliedFilterIds` prop?
    const selectedDialectFilter = selectn(`${facetType}`, this.props.routeParams)
    const prevSelectedDialectFilter = selectn(`${facetType}`, prevProps.routeParams)

    if (selectedDialectFilter !== prevSelectedDialectFilter) {
      const selectedParams = this.historyData[selectedDialectFilter]
      if (selectedParams) {
        this.setSelected(selectedParams)
        this.selectedDialectFilter = undefined
      }
    }
  }

  componentWillUnmount() {
    window.removeEventListener('popstate', this.handleHistoryEvent)
  }

  render() {
    return this.props.children({
      listItemData: this.generateListItemData(),
    })
  }

  // Creates url used with <Link>/<a> in Presentation layer
  generateUrlDialectFilterListItem = (filterId) => {
    const { type } = this.props
    let href = `/${this.props.splitWindowPath.join('/')}`
    if (type === 'words') {
      href = `/${appendPathArrayAfterLandmark({
        pathArray: ['categories', filterId],
        splitWindowPath: this.props.splitWindowPath,
        landmarkArray: ['words'],
      })}`
    }
    if (type === 'phrases') {
      href = `/${appendPathArrayAfterLandmark({
        pathArray: ['phrasebook', filterId],
        splitWindowPath: this.props.splitWindowPath,
        landmarkArray: ['learn'],
      })}`
    }
    return href
  }

  // generateListItemData
  // ---------------------------------------------------
  // Generates data structure representing the list
  // Processes `filters` and transforms into an object to be used by the Presentation layer
  //
  // returns [
  //   { // Top level category
  //     children: [ // Children categories of the top level category
  //       {
  //         hasActiveParent,
  //         href,
  //         isActive,
  //         text,
  //         uid,
  //       },
  //       /* More children... */
  //     ],
  //     hasActiveChild,
  //     href,
  //     isActive,
  //     text,
  //     uid,
  //   },
  //   /* More top level categories... */
  // ]
  generateListItemData = (filters = this.filtersSorted) => {
    const { appliedFilterIds } = this.props

    const listItemData = []
    filters.forEach((filter) => {
      const childData = []
      const uidParent = filter.uid

      // Process children
      const children = selectn('contextParameters.children.entries', filter)
      const parentIsActive = appliedFilterIds.includes(uidParent)
      let hasActiveChild = false

      if (children.length > 0) {
        children.forEach((filterChild) => {
          // TEMP FIX FOR: FW-1337: Filtering trashed children documents
          if (filterChild.isTrashed) return
          const uidChild = filterChild.uid
          const childIsActive = appliedFilterIds.includes(uidChild)
          // Set flag for parent processing
          hasActiveChild = childIsActive
          // Save child data
          childData.push({
            hasActiveParent: parentIsActive,
            href: this.generateUrlDialectFilterListItem(uidChild),
            isActive: childIsActive,
            text: filterChild.title,
            uid: uidChild,
          })
        })
      }

      // Process parent
      listItemData.push({
        children: childData,
        hasActiveChild: hasActiveChild,
        href: this.generateUrlDialectFilterListItem(uidParent),
        isActive: parentIsActive,
        text: filter.title,
        uid: uidParent,
      })
    })
    return listItemData
  }

  // This generates data used when the back/forward buttons are clicked.
  // Not certain why it's needed and if it is, why can't we reuse data created by `generateListItemData`?
  // Creates:
  // {
  //   [uid]: {
  //     checkedFacetUid, // ''
  //     childrenIds, // []
  //     href, // ''
  //     parentFacetUid, // '' || undefined
  //   },
  //   // more entries...
  // }
  generateDataForHistoryEvents = (filters) => {
    const newHistoryData = {}
    filters.forEach((filter) => {
      const uidParent = filter.uid

      // Process children
      const childrenUids = []
      const children = selectn('contextParameters.children.entries', filter)
      if (children.length > 0) {
        children.forEach((filterChild) => {
          // TEMP FIX FOR: FW-1337: Filtering trashed children documents
          if (filterChild.isTrashed) return
          const uidChild = filterChild.uid

          childrenUids.push(uidChild)

          // Saving for history events
          newHistoryData[uidChild] = {
            href: this.generateUrlDialectFilterListItem(uidChild),
            checkedFacetUid: uidChild,
            childrenIds: null,
            parentFacetUid: uidParent,
          }
        })
      }

      // Process parent
      const parentClickParams = {
        href: this.generateUrlDialectFilterListItem(uidParent),
        checkedFacetUid: uidParent,
        childrenIds: childrenUids,
        parentFacetUid: undefined,
      }
      newHistoryData[uidParent] = parentClickParams
    })

    return newHistoryData
  }

  setSelected = ({ href, checkedFacetUid, childrenIds }) => {
    // 'check' new
    this.setState(() => {
      const selected = {
        checkedFacetUid,
        childrenIds,
      }

      this.setDialectFilter({
        facetField: this.state.facetField,
        href,
        selected,
      })
    })
  }

  // Saves to redux so other components can respond (eg: "you are searching...")
  // also calls any passed in callback from the parent/layout component (typically used to update it's state)
  setDialectFilter = async ({ facetField, href, selected, updateUrl }) => {
    this.props.setDialectFilterCallback({
      facetField,
      selected,
      href,
      updateUrl,
    })
  }

  // `handleHistoryEvent` is called during history events
  handleHistoryEvent = () => {
    const _filterId = selectn(`routeParams.${this.props.facetType}`, this.props)
    if (_filterId) {
      const selectedParams = this.historyData[_filterId]
      if (selectedParams) {
        const { href, checkedFacetUid, childrenIds } = selectedParams
        this.setState(() => {
          const selected = {
            checkedFacetUid,
            childrenIds,
          }
          this.setDialectFilter({
            facetField: this.state.facetField,
            href,
            selected,
            updateUrl: false,
          })
        })
      }
    }
  }

  // Sort functions:
  sortByTitle = (a, b) => {
    if (a.title < b.title) return -1
    if (a.title > b.title) return 1
    return 0
  }

  sortDialectFilters = (filters = []) => {
    const _filters = [...filters]
    // Sort root level
    _filters.sort(this.sortByTitle)
    const _filtersSorted = _filters.map((filter) => {
      // Sort children
      const children = selectn('contextParameters.children.entries', filter)
      if (children.length > 0) {
        children.sort(this.sortByTitle)
      }
      return filter
    })
    return _filtersSorted
  }
}

// PROPTYPES
const { any, array, func, instanceOf, object, string, oneOf } = PropTypes
DialectFilterListData.propTypes = {
  // React built-in prop:
  children: any,
  // Props applied to instance of data component:
  appliedFilterIds: instanceOf(Set), // Selected list items, rename?
  path: string, // Used with facets
  setDialectFilterCallback: func,
  facets: array.isRequired,
  facetType: oneOf(['category', 'phraseBook']),
  type: oneOf(['words', 'phrases']).isRequired,
  workspaceKey: string, // Used with facetField
  // Redux props:
  // REDUX: reducers/state
  computeCategories: object.isRequired,
  routeParams: object.isRequired,
  splitWindowPath: array.isRequired,
  // REDUX: actions/dispatch/func
  fetchCategories: func.isRequired,
}
DialectFilterListData.defaultProps = {
  appliedFilterIds: new Set(),
  dialectFilterListWillUnmount: () => {},
}

// REDUX: reducers/state
const mapStateToProps = (state) => {
  const { fvCategory, navigation, windowPath } = state
  const { computeCategories } = fvCategory
  const { route } = navigation
  const { splitWindowPath } = windowPath
  return {
    computeCategories,
    routeParams: route.routeParams,
    splitWindowPath,
  }
}
// REDUX: actions/dispatch/func
const mapDispatchToProps = {
  fetchCategories,
}

export default connect(mapStateToProps, mapDispatchToProps)(DialectFilterListData)
