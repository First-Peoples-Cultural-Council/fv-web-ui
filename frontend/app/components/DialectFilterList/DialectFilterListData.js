import { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import selectn from 'selectn'
import useNavigationHelpers from 'common/useNavigationHelpers'
function DialectFilterListData({ children, filterListData, queryParam }) {
  const { getSearchAsObject } = useNavigationHelpers()
  const [listItemData, setListItemData] = useState([])
  const [sortedFacets, setSortedFacets] = useState()
  // Sort functions:
  const sortByTitle = (a, b) => {
    if (a.title < b.title) return -1
    if (a.title > b.title) return 1
    return 0
  }
  const sortDialectFilters = (filters = []) => {
    const _filters = [...filters]
    // Sort root level
    _filters.sort(sortByTitle)
    const _filtersSorted = _filters.map((filter) => {
      // Sort children
      const childrenEntries = selectn('contextParameters.children.entries', filter)
      if (childrenEntries.length > 0) {
        childrenEntries.sort(sortByTitle)
      }
      return filter
    })
    return _filtersSorted
  }
  useEffect(() => {
    setSortedFacets(sortDialectFilters(filterListData))
  }, [filterListData])

  const { category: queryCategory, phraseBook: queryPhraseBook } = getSearchAsObject()
  useEffect(() => {
    if (sortedFacets) {
      // Need to rebuild data in order to set the active flags within the data
      setListItemData(generateListItemData(sortedFacets))
    }
  }, [sortedFacets, queryCategory, queryPhraseBook])

  // generateListItemData
  // ---------------------------------------------------
  // Generates data structure representing the list
  // Processes `filters` and transforms into an object to be used by the Presentation layer
  //
  // returns = [
  //   { // Top level category
  //     children: [ // Children categories of the top level category
  //       {
  //         hasActiveParent,
  //         href,
  //         isActive,
  //         text,
  //         uid,
  //       },
  //       /* More children... */
  //     ],
  //     hasActiveChild,
  //     href,
  //     isActive,
  //     text,
  //     uid,
  //   },
  //   /* More top level categories... */
  // ]
  const generateListItemData = (filters) => {
    const toReturn = []
    const selectedFilterId = queryParam === 'category' ? queryCategory : queryPhraseBook
    filters.forEach((filter) => {
      const childData = []
      const uidParent = filter.uid

      // Process children
      const childrenEntries = selectn('contextParameters.children.entries', filter)
      const parentIsActive = selectedFilterId === uidParent
      let hasActiveChild = false

      if (childrenEntries.length > 0) {
        childrenEntries.forEach((filterChild) => {
          // TEMP FIX FOR: FW-1337: Filtering trashed children documents
          if (filterChild.isTrashed) return
          const uidChild = filterChild.uid
          const childIsActive = selectedFilterId === uidChild
          // Set flag for parent processing
          hasActiveChild = childIsActive
          // Save child data
          childData.push({
            hasActiveParent: parentIsActive,
            href: `${window.location.pathname}?${queryParam}=${uidChild}`,
            isActive: childIsActive,
            text: filterChild.title,
            uid: uidChild,
          })
        })
      }

      // Process parent
      toReturn.push({
        children: childData,
        hasActiveChild: hasActiveChild,
        href: `${window.location.pathname}?${queryParam}=${uidParent}`,
        isActive: parentIsActive,
        text: filter.title,
        uid: uidParent,
      })
    })
    return toReturn
  }
  return children({
    listItemData: listItemData,
  })
}

// PROPTYPES
const { array, func, oneOf } = PropTypes
DialectFilterListData.propTypes = {
  children: func,
  filterListData: array.isRequired,
  queryParam: oneOf(['category', 'phraseBook']),
}

export default DialectFilterListData
