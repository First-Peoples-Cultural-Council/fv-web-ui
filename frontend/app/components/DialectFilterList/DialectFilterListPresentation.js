import React from 'react'
import PropTypes from 'prop-types'
import Link from 'components/Link'
import '!style-loader!css-loader!./DialectFilterList.css'
function DialectFilterListPresentation({ title, listItemData }) {
  const generateList = () => {
    const listItems = listItemData.map((parent, parentIndex) => {
      const {
        isActive: parentIsActive,
        hasActiveChild: parentHasActiveChild,
        uid: parentUid,
        text: parentText,
        href: parentHref,
      } = parent
      const parentActiveClass = parentIsActive ? 'DialectFilterListLink--active' : ''

      const childrenNodes =
        parent.children.length > 0
          ? parent.children.map((child, childIndex) => {
              const { isActive: childIsActive, uid: childUid, text: childText, href: childHref } = child

              let childActiveClass = ''
              if (parentActiveClass) {
                childActiveClass = 'DialectFilterListLink--activeParent'
              }
              if (childIsActive) {
                childActiveClass = 'DialectFilterListLink--active'
              }
              return (
                <li key={childUid}>
                  <Link
                    dataTestId={`DialectFilterListItem-${parentIndex}-${childIndex}`}
                    className={`DialectFilterListLink DialectFilterListLink--child ${childActiveClass}`}
                    href={childHref}
                    title={childText}
                  >
                    {childText}
                  </Link>
                </li>
              )
            })
          : null

      const parentActiveChildClass = parentHasActiveChild ? 'DialectFilterListLink--activeChild' : ''
      return (
        <li key={parentUid} className={`DialectFilterListItemParent ${parentActiveClass} ${parentActiveChildClass}`}>
          <div className="DialectFilterListItemGroup">
            <Link
              dataTestId={`DialectFilterListItem-${parentIndex}`}
              className={`DialectFilterListLink DialectFilterListLink--parent ${parentActiveClass}`}
              href={parentHref}
              title={parentText}
            >
              {parentText}
            </Link>
          </div>
          {childrenNodes && <ul className="DialectFilterListList">{childrenNodes}</ul>}
        </li>
      )
    })

    return listItems ? <ul className="DialectFilterListList DialectFilterListList--root">{listItems}</ul> : null
  }

  return (
    <div className="DialectFilterList" data-testid="DialectFilterList">
      <h2>{title}</h2>
      {listItemData.length !== 0 && generateList()}
    </div>
  )
}

// Proptypes
const { array, string } = PropTypes
DialectFilterListPresentation.propTypes = {
  title: string.isRequired,
  listItemData: array,
}

DialectFilterListPresentation.defaultProps = {
  listItemData: [],
}

export default DialectFilterListPresentation
