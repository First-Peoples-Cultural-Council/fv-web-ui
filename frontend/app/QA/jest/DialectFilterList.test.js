beforeAll(() => jest.spyOn(React, 'useEffect').mockImplementation(React.useLayoutEffect))
afterAll(() => React.useEffect.mockRestore())

// Mock: react-redux
// --------------------------------------------------
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}))

// Mock: useRoute
// --------------------------------------------------
import * as useNavigationHelpers from 'common/useNavigationHelpers'
const spyUseNavigationHelpers = jest.spyOn(useNavigationHelpers, 'default')
spyUseNavigationHelpers.mockReturnValue({
  getSearchAsObject: () => {
    return { category: '2100b13d-218e-4660-927a-9f80475a8019' }
  },
})

// Let us begin!
// ==================================================
import React from 'react'
import renderer from 'react-test-renderer'
import { queryParam, filterListData } from './DialectFilterList.test.data'
import DialectFilterListData from 'components/DialectFilterList/DialectFilterListData'
import DialectFilterListPresentation from 'components/DialectFilterList/DialectFilterListPresentation'
import { Provider } from 'react-redux'
import store from 'state/store'
// import util from 'util'
describe('DialectFilterList', () => {
  // Updated in Data test > DialectFilterListData.children
  // Reused in Presentation
  let DataOutput = {}
  // Data
  test('Data', () => {
    renderer.create(
      <DialectFilterListData queryParam={queryParam} filterListData={filterListData}>
        {(dataLayerOutput) => {
          DataOutput = dataLayerOutput
          // console.log('util', util.inspect(output, {depth: null}))
          return null
        }}
      </DialectFilterListData>
    )
    expect(DataOutput).toMatchInlineSnapshot(`
      Object {
        "listItemData": Array [
          Object {
            "children": Array [
              Object {
                "hasActiveParent": true,
                "href": "/?category=2e6a992f-2023-4be9-8317-2e7a824a349b",
                "isActive": false,
                "text": "Child Category IIII",
                "uid": "2e6a992f-2023-4be9-8317-2e7a824a349b",
              },
              Object {
                "hasActiveParent": true,
                "href": "/?category=727cbb38-913f-40d9-9bde-ad83bf4129c9",
                "isActive": false,
                "text": "Child Category IIᑕ",
                "uid": "727cbb38-913f-40d9-9bde-ad83bf4129c9",
              },
            ],
            "hasActiveChild": false,
            "href": "/?category=2100b13d-218e-4660-927a-9f80475a8019",
            "isActive": true,
            "text": "Parent Category II",
            "uid": "2100b13d-218e-4660-927a-9f80475a8019",
          },
          Object {
            "children": Array [
              Object {
                "hasActiveParent": false,
                "href": "/?category=2446869b-8109-42ff-89a6-b37b4ddb35c8",
                "isActive": false,
                "text": "Child Category OII",
                "uid": "2446869b-8109-42ff-89a6-b37b4ddb35c8",
              },
              Object {
                "hasActiveParent": false,
                "href": "/?category=15bb8a84-f203-4a64-aa37-b97787d8b1ca",
                "isActive": false,
                "text": "Child Category Oᑕ",
                "uid": "15bb8a84-f203-4a64-aa37-b97787d8b1ca",
              },
            ],
            "hasActiveChild": false,
            "href": "/?category=b64f12e3-3f83-4046-99d5-872ff16e0762",
            "isActive": false,
            "text": "Parent Category O",
            "uid": "b64f12e3-3f83-4046-99d5-872ff16e0762",
          },
          Object {
            "children": Array [
              Object {
                "hasActiveParent": false,
                "href": "/?category=c2d3c4c2-b35c-438e-ae87-872fefe274ab",
                "isActive": false,
                "text": "Child Category ᑕII",
                "uid": "c2d3c4c2-b35c-438e-ae87-872fefe274ab",
              },
              Object {
                "hasActiveParent": false,
                "href": "/?category=9c5c2e23-25df-4faa-bdb6-b297ba105461",
                "isActive": false,
                "text": "Child Category ᑕᑕ",
                "uid": "9c5c2e23-25df-4faa-bdb6-b297ba105461",
              },
            ],
            "hasActiveChild": false,
            "href": "/?category=db456452-da58-4c75-81ba-8c5d13153c4e",
            "isActive": false,
            "text": "Parent Category ᑕ",
            "uid": "db456452-da58-4c75-81ba-8c5d13153c4e",
          },
        ],
      }
    `)
  })
  // Presentation
  test('Presentation', () => {
    const component = renderer.create(
      <Provider store={store}>
        <DialectFilterListPresentation title="Jest Testing" listItemData={DataOutput.listItemData} />
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
