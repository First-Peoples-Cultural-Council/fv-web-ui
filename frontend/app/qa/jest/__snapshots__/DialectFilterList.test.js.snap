// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialectFilterList Presentation 1`] = `
<div
  className="DialectFilterList"
  data-testid="DialectFilterList"
>
  <h2>
    Jest Testing
  </h2>
  <ul
    className="DialectFilterListList DialectFilterListList--root"
  >
    <li
      className="DialectFilterListItemParent DialectFilterListLink--active "
    >
      <div
        className="DialectFilterListItemGroup"
      >
        <a
          className="DialectFilterListLink DialectFilterListLink--parent DialectFilterListLink--active"
          data-testid="DialectFilterListItem-0"
          href="/?category=2100b13d-218e-4660-927a-9f80475a8019"
          onClick={[Function]}
        >
          Parent Category II
        </a>
      </div>
      <ul
        className="DialectFilterListList"
      >
        <li>
          <a
            className="DialectFilterListLink DialectFilterListLink--child DialectFilterListLink--activeParent"
            data-testid="DialectFilterListItem-0-0"
            href="/?category=2e6a992f-2023-4be9-8317-2e7a824a349b"
            onClick={[Function]}
          >
            Child Category IIII
          </a>
        </li>
        <li>
          <a
            className="DialectFilterListLink DialectFilterListLink--child DialectFilterListLink--activeParent"
            data-testid="DialectFilterListItem-0-1"
            href="/?category=727cbb38-913f-40d9-9bde-ad83bf4129c9"
            onClick={[Function]}
          >
            Child Category IIᑕ
          </a>
        </li>
      </ul>
    </li>
    <li
      className="DialectFilterListItemParent  "
    >
      <div
        className="DialectFilterListItemGroup"
      >
        <a
          className="DialectFilterListLink DialectFilterListLink--parent "
          data-testid="DialectFilterListItem-1"
          href="/?category=b64f12e3-3f83-4046-99d5-872ff16e0762"
          onClick={[Function]}
        >
          Parent Category O
        </a>
      </div>
      <ul
        className="DialectFilterListList"
      >
        <li>
          <a
            className="DialectFilterListLink DialectFilterListLink--child "
            data-testid="DialectFilterListItem-1-0"
            href="/?category=2446869b-8109-42ff-89a6-b37b4ddb35c8"
            onClick={[Function]}
          >
            Child Category OII
          </a>
        </li>
        <li>
          <a
            className="DialectFilterListLink DialectFilterListLink--child "
            data-testid="DialectFilterListItem-1-1"
            href="/?category=15bb8a84-f203-4a64-aa37-b97787d8b1ca"
            onClick={[Function]}
          >
            Child Category Oᑕ
          </a>
        </li>
      </ul>
    </li>
    <li
      className="DialectFilterListItemParent  "
    >
      <div
        className="DialectFilterListItemGroup"
      >
        <a
          className="DialectFilterListLink DialectFilterListLink--parent "
          data-testid="DialectFilterListItem-2"
          href="/?category=db456452-da58-4c75-81ba-8c5d13153c4e"
          onClick={[Function]}
        >
          Parent Category ᑕ
        </a>
      </div>
      <ul
        className="DialectFilterListList"
      >
        <li>
          <a
            className="DialectFilterListLink DialectFilterListLink--child "
            data-testid="DialectFilterListItem-2-0"
            href="/?category=c2d3c4c2-b35c-438e-ae87-872fefe274ab"
            onClick={[Function]}
          >
            Child Category ᑕII
          </a>
        </li>
        <li>
          <a
            className="DialectFilterListLink DialectFilterListLink--child "
            data-testid="DialectFilterListItem-2-1"
            href="/?category=9c5c2e23-25df-4faa-bdb6-b297ba105461"
            onClick={[Function]}
          >
            Child Category ᑕᑕ
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
